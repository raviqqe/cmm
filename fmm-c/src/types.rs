use crate::names::*;
use fmm::types::{self, Type};
use std::collections::FnvHashMap;

pub fn compile_typed_name(type_: &Type, name: &str, type_ids: &FnvHashMap<Type, String>) -> String {
    match type_ {
        Type::Function(function) => {
            compile_function_name(function, &format!("(*{})", name), type_ids)
        }
        Type::Primitive(primitive) => compile_primitive_type_id(*primitive) + " " + name,
        Type::Record(record) => compile_record_type_id(record, type_ids) + " " + name,
        Type::Pointer(pointer) => {
            compile_typed_name(pointer.element(), &format!("*{}", name), type_ids)
        }
        Type::Union(union) => compile_union_type_id(union, type_ids) + " " + name,
    }
}

pub fn compile_function_name(
    function: &types::Function,
    name: &str,
    type_ids: &FnvHashMap<Type, String>,
) -> String {
    compile_typed_name(
        function.result(),
        &format!(
            "{}({})",
            name,
            function
                .arguments()
                .iter()
                .enumerate()
                .map(|(index, type_)| compile_typed_name(
                    type_,
                    &generate_argument_name(index),
                    type_ids
                ))
                .collect::<Vec<_>>()
                .join(",")
        ),
        type_ids,
    )
}

pub fn compile_type_id(type_: &Type, type_ids: &FnvHashMap<Type, String>) -> String {
    match type_ {
        Type::Function(function) => compile_function_name(function, "(*)", type_ids),
        Type::Primitive(primitive) => compile_primitive_type_id(*primitive),
        Type::Record(record) => compile_record_type_id(record, type_ids),
        Type::Pointer(pointer) => compile_typed_name(pointer.element(), "*", type_ids),
        Type::Union(union) => compile_union_type_id(union, type_ids),
    }
}

pub fn compile_atomic_pointer_type_id(type_: &Type, type_ids: &FnvHashMap<Type, String>) -> String {
    match type_ {
        Type::Function(function) => compile_function_name(function, "(*_Atomic *)", type_ids),
        Type::Pointer(pointer) => compile_typed_name(pointer.element(), "_Atomic *", type_ids),
        Type::Primitive(_) | Type::Record(_) | Type::Union(_) => {
            "_Atomic ".to_owned() + &compile_type_id(type_, type_ids) + " *"
        }
    }
}

pub fn compile_primitive_type_id(primitive: types::Primitive) -> String {
    match primitive {
        types::Primitive::Boolean => "bool",
        types::Primitive::Float32 => "float",
        types::Primitive::Float64 => "double",
        types::Primitive::Integer8 => "uint8_t",
        types::Primitive::Integer32 => "uint32_t",
        types::Primitive::Integer64 => "uint64_t",
        types::Primitive::PointerInteger => "size_t",
    }
    .into()
}

pub fn compile_record_type_id(record: &types::Record, type_ids: &FnvHashMap<Type, String>) -> String {
    "struct ".to_owned() + &type_ids[&record.clone().into()]
}

pub fn compile_union_type_id(union: &types::Union, type_ids: &FnvHashMap<Type, String>) -> String {
    "union ".to_owned() + &type_ids[&union.clone().into()]
}

pub fn compile_record_fields(record: &types::Record, type_ids: &FnvHashMap<Type, String>) -> String {
    record
        .fields()
        .iter()
        .enumerate()
        .map(|(index, type_)| {
            compile_typed_name(type_, &generate_record_field_name(index), type_ids) + ";"
        })
        .collect::<Vec<_>>()
        .join("")
}

pub fn compile_union_members(union: &types::Union, type_ids: &FnvHashMap<Type, String>) -> String {
    union
        .members()
        .iter()
        .enumerate()
        .map(|(index, type_)| {
            compile_typed_name(type_, &generate_union_member_name(index), type_ids) + ";"
        })
        .collect::<Vec<_>>()
        .join("")
}
