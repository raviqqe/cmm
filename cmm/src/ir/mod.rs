mod allocate_heap;
mod allocate_stack;
mod alternative;
mod argument;
mod arithmetic_operation;
mod assignment;
mod atomic_load;
mod atomic_store;
mod bitcast;
mod call;
mod compare_and_swap;
mod comparison_operation;
mod deconstruct_record;
mod default_alternative;
mod expression;
mod function_declaration;
mod function_definition;
mod if_;
mod instruction;
mod load;
mod module;
mod pointer_address;
mod primitive;
mod record;
mod record_address;
mod return_;
mod store;
mod switch;
mod union;
mod union_address;
mod variable;
mod variable_declaration;
mod variable_definition;

pub use allocate_heap::*;
pub use allocate_stack::*;
pub use alternative::*;
pub use argument::*;
pub use arithmetic_operation::*;
pub use assignment::*;
pub use atomic_load::*;
pub use atomic_store::*;
pub use bitcast::*;
pub use call::*;
pub use compare_and_swap::*;
pub use comparison_operation::*;
pub use deconstruct_record::*;
pub use default_alternative::*;
pub use expression::*;
pub use function_declaration::*;
pub use function_definition::*;
pub use if_::*;
pub use instruction::*;
pub use load::*;
pub use module::*;
pub use pointer_address::*;
pub use primitive::*;
pub use record::*;
pub use record_address::*;
pub use return_::*;
pub use store::*;
pub use switch::*;
pub use union::*;
pub use union_address::*;
pub use variable::*;
pub use variable_declaration::*;
pub use variable_definition::*;
