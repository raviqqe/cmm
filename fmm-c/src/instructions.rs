use crate::{expressions::*, names::*, types::*};
use fmm::{
    ir::*,
    types::{self, GENERIC_POINTER_TYPE},
};
use std::collections::{HashMap, HashSet};

pub fn compile_block(
    block: &Block,
    branch_variable_name: Option<&str>,
    global_variables: &HashSet<String>,
    type_ids: &HashMap<fmm::types::Type, String>,
) -> String {
    block
        .instructions()
        .iter()
        .map(|instruction| compile_instruction(instruction, global_variables, type_ids))
        .chain(vec![compile_terminal_instruction(
            block.terminal_instruction(),
            branch_variable_name,
            global_variables,
            type_ids,
        )])
        .map(|string| "".to_owned() + &string)
        .collect::<Vec<_>>()
        .join("\n")
}

fn compile_instruction(
    instruction: &Instruction,
    global_variables: &HashSet<String>,
    type_ids: &HashMap<fmm::types::Type, String>,
) -> String {
    let compile_expression =
        |expression| compile_expression(expression, global_variables, type_ids);
    let compile_typed_name = |type_, name| compile_typed_name(type_, name, type_ids);
    let compile_type_id = |type_| compile_type_id(type_, type_ids);

    match instruction {
        Instruction::AllocateHeap(allocate) => {
            format!(
                "{}=malloc({});",
                compile_typed_name(&GENERIC_POINTER_TYPE, allocate.name()),
                compile_expression(allocate.size()),
            )
        }
        Instruction::AllocateStack(allocate) => {
            let entity_name = allocate.name().to_owned() + "_entity";

            format!(
                "{};{}=&{};",
                compile_typed_name(allocate.type_(), &entity_name),
                compile_typed_name(
                    &types::Pointer::new(allocate.type_().clone()).into(),
                    allocate.name(),
                ),
                entity_name,
            )
        }
        Instruction::AtomicLoad(load) => format!(
            "{}=({})atomic_load_explicit(({}){},{});",
            compile_typed_name(load.type_(), load.name()),
            compile_type_id(load.type_()),
            compile_atomic_pointer_type_id(load.type_(), type_ids),
            compile_expression(load.pointer()),
            compile_atomic_ordering(load.ordering()),
        ),
        Instruction::AtomicOperation(operation) => format!(
            "{}=atomic_fetch_{}_explicit(({}){},{},{});",
            compile_typed_name(&operation.type_().into(), operation.name()),
            match operation.operator() {
                AtomicOperator::Add => "add",
                AtomicOperator::Subtract => "sub",
            },
            compile_atomic_pointer_type_id(&operation.type_().into(), type_ids),
            compile_expression(operation.pointer()),
            compile_expression(operation.value()),
            compile_atomic_ordering(operation.ordering()),
        ),
        Instruction::AtomicStore(store) => format!(
            "atomic_store_explicit(({}){},{},{});",
            compile_atomic_pointer_type_id(store.type_(), type_ids),
            compile_expression(store.pointer()),
            compile_expression(store.value()),
            compile_atomic_ordering(store.ordering()),
        ),
        Instruction::Call(call) => format!(
            "{}={}({});",
            compile_typed_name(call.type_().result(), call.name()),
            compile_expression(call.function()),
            call.arguments()
                .iter()
                .map(compile_expression)
                .collect::<Vec<_>>()
                .join(",")
        ),
        Instruction::CompareAndSwap(cas) => {
            let name = "_cas_".to_owned() + cas.name();

            format!(
                "{}={};bool {}=atomic_compare_exchange_strong_explicit(({}){},&{},{},{},{});",
                compile_typed_name(cas.type_(), &name),
                compile_expression(cas.old_value()),
                cas.name(),
                compile_atomic_pointer_type_id(cas.type_(), type_ids),
                compile_expression(cas.pointer()),
                name,
                compile_expression(cas.new_value()),
                compile_atomic_ordering(cas.success_ordering()),
                compile_atomic_ordering(cas.failure_ordering()),
            )
        }
        Instruction::DeconstructRecord(deconstruct) => format!(
            "{}={}.{};",
            compile_typed_name(
                &deconstruct.type_().fields()[deconstruct.element_index()],
                deconstruct.name(),
            ),
            compile_expression(deconstruct.record()),
            generate_record_field_name(deconstruct.element_index()),
        ),
        Instruction::DeconstructUnion(deconstruct) => format!(
            "{}={}.{};",
            compile_typed_name(
                &deconstruct.type_().members()[deconstruct.member_index()],
                deconstruct.name(),
            ),
            compile_expression(deconstruct.union()),
            generate_union_member_name(deconstruct.member_index()),
        ),
        Instruction::Fence(fence) => {
            format!(
                "atomic_thread_fence({});",
                compile_atomic_ordering(fence.ordering())
            )
        }
        Instruction::FreeHeap(free) => {
            format!("free((void *)({}));", compile_expression(free.pointer()))
        }
        Instruction::If(if_) => {
            let compile_block =
                |block| compile_block(block, Some(if_.name()), global_variables, type_ids);

            format!(
                "{};if({}){{\n{}\n}}else{{\n{}\n}}",
                compile_typed_name(&if_.type_().clone(), if_.name()),
                compile_expression(if_.condition()),
                compile_block(if_.then()),
                compile_block(if_.else_())
            )
        }
        Instruction::Load(load) => format!(
            "{}=*{};",
            compile_typed_name(load.type_(), load.name()),
            compile_expression(load.pointer()),
        ),
        Instruction::PassThrough(pass) => format!(
            "{}={};",
            compile_typed_name(pass.type_(), pass.name()),
            compile_expression(pass.expression()),
        ),
        Instruction::ReallocateHeap(reallocate) => {
            format!(
                "{}=realloc({},{});",
                compile_typed_name(&types::GENERIC_POINTER_TYPE.clone(), reallocate.name()),
                compile_expression(reallocate.pointer()),
                compile_expression(reallocate.size()),
            )
        }
        Instruction::Store(store) => format!(
            "*{}={};",
            compile_expression(store.pointer()),
            compile_expression(store.value()),
        ),
    }
}

fn compile_terminal_instruction(
    instruction: &TerminalInstruction,
    block_variable_name: Option<&str>,
    global_variables: &HashSet<String>,
    type_ids: &HashMap<fmm::types::Type, String>,
) -> String {
    let compile_expression =
        |expression| compile_expression(expression, global_variables, type_ids);

    match instruction {
        TerminalInstruction::Branch(branch) => block_variable_name
            .map(|name| format!("{}={};", name, compile_expression(branch.expression(),)))
            .unwrap_or_default(),
        TerminalInstruction::Return(return_) => {
            format!("return {};", compile_expression(return_.expression(),))
        }
        TerminalInstruction::Unreachable => "abort();".into(),
    }
}

fn compile_atomic_ordering(ordering: AtomicOrdering) -> String {
    format!(
        "memory_order_{}",
        match ordering {
            AtomicOrdering::Relaxed => "relaxed",
            AtomicOrdering::Release => "release",
            AtomicOrdering::Acquire => "acquire",
            AtomicOrdering::AcquireRelease => "acq_rel",
            AtomicOrdering::SequentiallyConsistent => "seq_cst",
        }
    )
}
