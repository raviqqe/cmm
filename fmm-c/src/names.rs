pub fn generate_argument_name(index: usize) -> String {
    format!("a{}", index)
}

pub fn generate_record_field_name(index: usize) -> String {
    format!("e{}", index)
}

pub fn generate_union_member_name(index: usize) -> String {
    format!("m{}", index)
}

pub fn generate_record_type_name(index: usize) -> String {
    format!("r{}", index)
}

pub fn generate_union_type_name(index: usize) -> String {
    format!("u{}", index)
}
