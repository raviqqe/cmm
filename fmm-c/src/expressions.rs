use super::types::*;
use crate::names::*;
use fmm::ir::*;
use fmm::types;

pub fn compile_expression(expression: &Expression) -> String {
    match expression {
        Expression::Primitive(primitive) => compile_primitive(*primitive),
        Expression::Record(record) => {
            format!(
                "({}){{{}}}",
                compile_record_type_id(record.type_()),
                record
                    .elements()
                    .iter()
                    .map(compile_expression)
                    .collect::<Vec<_>>()
                    .join(",")
            )
        }
        Expression::Undefined(undefined) => compile_undefined(undefined),
        Expression::Union(union) => {
            format!(
                "({}){{.{}={}}}",
                compile_union_type_id(union.type_()),
                generate_union_member_name(union.member_index()),
                compile_expression(union.member())
            )
        }
        Expression::Variable(variable) => variable.name().into(),
    }
}

fn compile_undefined(undefined: &Undefined) -> String {
    match undefined.type_() {
        types::Type::Function(_) => "NULL".into(),
        types::Type::Primitive(primitive) => compile_undefined_primitive(*primitive).into(),
        types::Type::Pointer(_) => "NULL".into(),
        types::Type::Record(record) => format!("({}){{}}", compile_record_type_id(record)),
        types::Type::Union(union) => format!("({}){{}}", compile_union_type_id(union)),
    }
}

fn compile_primitive(primitive: Primitive) -> String {
    match primitive {
        Primitive::Boolean(boolean) => format!("{}", boolean),
        Primitive::Float32(number) => format!("{}", number),
        Primitive::Float64(number) => format!("{}", number),
        Primitive::Integer8(number) => format!("{}", number),
        Primitive::Integer32(number) => format!("{}", number),
        Primitive::Integer64(number) => format!("{}", number),
        Primitive::PointerInteger(number) => format!("{}", number),
    }
}

fn compile_undefined_primitive(primitive: types::Primitive) -> &'static str {
    match primitive {
        types::Primitive::Boolean => "false",
        types::Primitive::Float32 | types::Primitive::Float64 => "0.0",
        types::Primitive::Integer8
        | types::Primitive::Integer32
        | types::Primitive::Integer64
        | types::Primitive::PointerInteger => "0",
    }
}
