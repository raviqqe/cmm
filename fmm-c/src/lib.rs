mod error;
mod expressions;
mod instructions;
mod names;
mod types;

pub use error::*;
use expressions::*;
use fmm::analysis::*;
use fmm::ir::*;
use instructions::*;
use names::*;
use types::*;

const INCLUDES: &[&str] = &[
    "#include <stdatomic.h>",
    "#include <stdbool.h>",
    "#include <stdint.h>",
    "#include <stdlib.h>",
];

pub fn compile(module: &Module) -> String {
    check_types(module).unwrap();

    let types = collect_types(module);

    let strings = INCLUDES
        .iter()
        .map(|&string| string.into())
        .chain(
            collect_record_types(&types)
                .iter()
                .map(compile_record_type_definition),
        )
        .chain(
            collect_union_types(&types)
                .iter()
                .map(compile_union_type_definition),
        )
        .chain(
            module
                .variable_declarations()
                .iter()
                .map(compile_variable_declaration),
        )
        .chain(
            module
                .variable_definitions()
                .iter()
                .map(compile_variable_forward_declaration),
        )
        .chain(
            module
                .function_declarations()
                .iter()
                .map(compile_function_declaration),
        )
        .chain(
            module
                .function_definitions()
                .iter()
                .map(compile_function_forward_declaration),
        )
        .chain(
            module
                .variable_definitions()
                .iter()
                .map(compile_variable_definition),
        )
        .chain(
            module
                .function_definitions()
                .iter()
                .map(compile_function_definition),
        )
        .collect::<Vec<_>>();

    strings
        .iter()
        .map(|string| string.as_str())
        .collect::<Vec<_>>()
        .join("\n")
}

fn compile_record_type_definition(record: &fmm::types::Record) -> String {
    format!(
        "struct {} {{{}}};",
        generate_record_type_name(record),
        compile_record_elements(record)
    )
}

fn compile_union_type_definition(union: &fmm::types::Union) -> String {
    format!(
        "union {} {{{}}};",
        generate_union_type_name(union),
        compile_union_members(union)
    )
}

fn compile_variable_declaration(declaration: &VariableDeclaration) -> String {
    "extern ".to_owned() + &compile_typed_name(declaration.type_(), declaration.name()) + ";"
}

fn compile_variable_forward_declaration(definition: &VariableDefinition) -> String {
    compile_variable_definition_lhs(definition) + ";"
}

fn compile_function_declaration(declaration: &FunctionDeclaration) -> String {
    "extern ".to_owned() + &compile_function_name(declaration.type_(), declaration.name()) + ";"
}

fn compile_function_forward_declaration(definition: &FunctionDefinition) -> String {
    if definition.is_global() {
        ""
    } else {
        "static "
    }
    .to_owned()
        + &compile_function_name(definition.type_(), definition.name())
        + ";"
}

fn compile_variable_definition(definition: &VariableDefinition) -> String {
    let entity_name = format!("{}_entity", definition.name());

    "static ".to_owned()
        + &compile_typed_name(
            definition.type_(),
            &(if definition.is_mutable() {
                ""
            } else {
                "const "
            }
            .to_owned()
                + &entity_name),
        )
        + " = "
        + &compile_expression(definition.body())
        + ";\n"
        + &compile_variable_definition_lhs(definition)
        + &format!(" = &{};", entity_name)
}

fn compile_variable_definition_lhs(definition: &VariableDefinition) -> String {
    if definition.is_global() {
        ""
    } else {
        "static "
    }
    .to_owned()
        + &compile_typed_name(
            definition.type_(),
            &(if definition.is_mutable() {
                ""
            } else {
                "const "
            }
            .to_owned()
                + "*const "
                + definition.name()),
        )
}

fn compile_function_definition(definition: &FunctionDefinition) -> String {
    if definition.is_global() {
        ""
    } else {
        "static "
    }
    .to_owned()
        + &compile_typed_name(
            definition.result_type(),
            &format!(
                "{}({})",
                definition.name(),
                definition
                    .arguments()
                    .iter()
                    .map(|argument| compile_typed_name(argument.type_(), argument.name()))
                    .collect::<Vec<_>>()
                    .join(",")
            ),
        )
        + "{\n"
        + &compile_block(definition.body(), None)
        + "\n}"
}

#[cfg(test)]
mod tests {
    use super::*;
    use fmm::types;

    fn compile_module(module: &Module) {
        let directory = tempfile::tempdir().unwrap();
        let file_path = directory.path().join("foo.c");
        let source = compile(module);

        println!("{}", source);

        std::fs::write(&file_path, source).unwrap();
        let output = std::process::Command::new("clang")
            .arg("-o")
            .arg(directory.path().join("foo.o"))
            .arg("-c")
            .arg(&file_path)
            .output()
            .unwrap();

        assert_eq!(String::from_utf8_lossy(&output.stdout), "");
        assert_eq!(String::from_utf8_lossy(&output.stderr), "");
        assert!(output.status.success());
    }

    #[test]
    fn compile_empty_module() {
        compile_module(&Module::new(vec![], vec![], vec![], vec![]));
    }

    mod variable_declarations {
        use super::*;

        #[test]
        fn compile_pointer_integer() {
            compile_module(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Primitive::PointerInteger,
                )],
                vec![],
                vec![],
                vec![],
            ));
        }

        #[test]
        fn compile_pointer_integer_pointer() {
            compile_module(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Pointer::new(types::Primitive::PointerInteger),
                )],
                vec![],
                vec![],
                vec![],
            ));
        }

        #[test]
        fn compile_function_pointer() {
            compile_module(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Function::new(vec![], types::Primitive::PointerInteger),
                )],
                vec![],
                vec![],
                vec![],
            ));
        }
    }

    mod function_declarations {
        use super::*;

        #[test]
        fn compile_function_pointer() {
            compile_module(&Module::new(
                vec![],
                vec![FunctionDeclaration::new(
                    "x",
                    types::Function::new(vec![], types::Primitive::PointerInteger),
                )],
                vec![],
                vec![],
            ));
        }
    }

    mod type_definitions {
        use super::*;

        #[test]
        fn compile_record_type_definition() {
            compile_module(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Record::new(vec![types::Primitive::PointerInteger.into()]),
                )],
                vec![],
                vec![],
                vec![],
            ));
        }

        #[test]
        fn compile_union_type_definition() {
            compile_module(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Union::new(vec![types::Primitive::PointerInteger.into()]),
                )],
                vec![],
                vec![],
                vec![],
            ));
        }
    }

    mod variable_definitions {
        use super::*;

        #[test]
        fn compile_constant_variable() {
            compile_module(&Module::new(
                vec![],
                vec![],
                vec![VariableDefinition::new(
                    "x",
                    fmm::ir::Primitive::PointerInteger(0),
                    types::Primitive::PointerInteger,
                    false,
                    true,
                )],
                vec![],
            ));
        }

        #[test]
        fn compile_mutable_variable() {
            compile_module(&Module::new(
                vec![],
                vec![],
                vec![VariableDefinition::new(
                    "x",
                    fmm::ir::Primitive::PointerInteger(0),
                    types::Primitive::PointerInteger,
                    true,
                    true,
                )],
                vec![],
            ));
        }

        #[test]
        fn compile_local_variable() {
            compile_module(&Module::new(
                vec![],
                vec![],
                vec![VariableDefinition::new(
                    "x",
                    fmm::ir::Primitive::PointerInteger(0),
                    types::Primitive::PointerInteger,
                    true,
                    false,
                )],
                vec![],
            ));
        }
    }

    mod function_definitions {
        use super::*;

        #[test]
        fn compile_global_function() {
            compile_module(&Module::new(
                vec![],
                vec![],
                vec![],
                vec![FunctionDefinition::new(
                    "x",
                    vec![],
                    fmm::ir::Block::new(
                        vec![],
                        fmm::ir::Return::new(
                            types::Primitive::PointerInteger,
                            fmm::ir::Primitive::PointerInteger(0),
                        ),
                    ),
                    types::Primitive::PointerInteger,
                    true,
                )],
            ));
        }

        #[test]
        fn compile_local_function() {
            compile_module(&Module::new(
                vec![],
                vec![],
                vec![],
                vec![FunctionDefinition::new(
                    "x",
                    vec![],
                    fmm::ir::Block::new(
                        vec![],
                        fmm::ir::Return::new(
                            types::Primitive::PointerInteger,
                            fmm::ir::Primitive::PointerInteger(0),
                        ),
                    ),
                    types::Primitive::PointerInteger,
                    false,
                )],
            ));
        }
    }

    mod instructions {
        use super::*;

        fn compile_function_definition(definition: FunctionDefinition) {
            compile_module(&Module::new(vec![], vec![], vec![], vec![definition]));
        }

        #[test]
        fn compile_unreachable() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(vec![], TerminalInstruction::Unreachable),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_allocate_heap() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![AllocateHeap::new(types::Primitive::PointerInteger, "y").into()],
                    Return::new(
                        types::Pointer::new(types::Primitive::PointerInteger),
                        Variable::new("y"),
                    ),
                ),
                types::Pointer::new(types::Primitive::PointerInteger),
                true,
            ));
        }

        #[test]
        fn compile_allocate_heap_with_function_pointer() {
            let function_type = types::Function::new(
                vec![types::Primitive::PointerInteger.into()],
                types::Primitive::PointerInteger,
            );

            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![AllocateHeap::new(function_type.clone(), "y").into()],
                    Return::new(
                        types::Pointer::new(function_type.clone()),
                        Variable::new("y"),
                    ),
                ),
                types::Pointer::new(function_type),
                true,
            ));
        }

        #[test]
        fn compile_arithmetic_operation() {
            for &operator in &[
                ArithmeticOperator::Add,
                ArithmeticOperator::Subtract,
                ArithmeticOperator::Multiply,
                ArithmeticOperator::Divide,
            ] {
                compile_function_definition(FunctionDefinition::new(
                    "f",
                    vec![],
                    Block::new(
                        vec![ArithmeticOperation::new(
                            types::Primitive::PointerInteger,
                            operator,
                            Primitive::PointerInteger(1),
                            Primitive::PointerInteger(1),
                            "x",
                        )
                        .into()],
                        Return::new(types::Primitive::PointerInteger, Variable::new("x")),
                    ),
                    types::Primitive::PointerInteger,
                    true,
                ));
            }
        }

        #[test]
        fn compile_comparison_operation() {
            for &operator in &[
                ComparisonOperator::Equal,
                ComparisonOperator::NotEqual,
                ComparisonOperator::LessThan,
                ComparisonOperator::GreaterThan,
                ComparisonOperator::LessThanOrEqual,
                ComparisonOperator::GreaterThanOrEqual,
            ] {
                compile_function_definition(FunctionDefinition::new(
                    "f",
                    vec![],
                    Block::new(
                        vec![ComparisonOperation::new(
                            types::Primitive::PointerInteger,
                            operator,
                            Primitive::PointerInteger(1),
                            Primitive::PointerInteger(1),
                            "x",
                        )
                        .into()],
                        Return::new(types::Primitive::Bool, Variable::new("x")),
                    ),
                    types::Primitive::Bool,
                    true,
                ));
            }
        }

        #[test]
        fn compile_atomic_load() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![Argument::new(
                    "x",
                    types::Pointer::new(types::Primitive::PointerInteger),
                )],
                Block::new(
                    vec![AtomicLoad::new(
                        types::Primitive::PointerInteger,
                        Variable::new("x"),
                        "y",
                    )
                    .into()],
                    Return::new(types::Primitive::PointerInteger, Variable::new("y")),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_atomic_load_with_function_pointer() {
            let function_type = types::Function::new(
                vec![types::Primitive::PointerInteger.into()],
                types::Primitive::PointerInteger,
            );

            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![Argument::new(
                    "x",
                    types::Pointer::new(function_type.clone()),
                )],
                Block::new(
                    vec![AtomicLoad::new(function_type.clone(), Variable::new("x"), "y").into()],
                    Return::new(function_type.clone(), Variable::new("y")),
                ),
                function_type,
                true,
            ));
        }

        #[test]
        fn compile_atomic_store() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![Argument::new(
                    "x",
                    types::Pointer::new(types::Primitive::PointerInteger),
                )],
                Block::new(
                    vec![AtomicStore::new(
                        types::Primitive::PointerInteger,
                        Undefined::new(types::Primitive::PointerInteger),
                        Variable::new("x"),
                    )
                    .into()],
                    Return::new(
                        types::Primitive::PointerInteger,
                        Primitive::PointerInteger(42),
                    ),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_atomic_store_with_function_pointer() {
            let function_type = types::Function::new(
                vec![types::Primitive::PointerInteger.into()],
                types::Primitive::PointerInteger,
            );

            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![Argument::new(
                    "x",
                    types::Pointer::new(function_type.clone()),
                )],
                Block::new(
                    vec![AtomicStore::new(
                        function_type.clone(),
                        Undefined::new(function_type),
                        Variable::new("x"),
                    )
                    .into()],
                    Return::new(
                        types::Primitive::PointerInteger,
                        Primitive::PointerInteger(42),
                    ),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_if() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![If::new(
                        types::Primitive::PointerInteger,
                        Primitive::Bool(true),
                        Block::new(
                            vec![],
                            Branch::new(
                                types::Primitive::PointerInteger,
                                Primitive::PointerInteger(42),
                            ),
                        ),
                        Block::new(
                            vec![],
                            Branch::new(
                                types::Primitive::PointerInteger,
                                Primitive::PointerInteger(42),
                            ),
                        ),
                        "x",
                    )
                    .into()],
                    Return::new(types::Primitive::PointerInteger, Variable::new("x")),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_if_with_return() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![If::new(
                        types::Primitive::PointerInteger,
                        Primitive::Bool(true),
                        Block::new(
                            vec![],
                            Return::new(
                                types::Primitive::PointerInteger,
                                Primitive::PointerInteger(42),
                            ),
                        ),
                        Block::new(
                            vec![],
                            Branch::new(
                                types::Primitive::PointerInteger,
                                Primitive::PointerInteger(42),
                            ),
                        ),
                        "x",
                    )
                    .into()],
                    Return::new(types::Primitive::PointerInteger, Variable::new("x")),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_if_with_unreachable() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![If::new(
                        types::Primitive::PointerInteger,
                        Primitive::Bool(true),
                        Block::new(vec![], TerminalInstruction::Unreachable),
                        Block::new(
                            vec![],
                            Branch::new(
                                types::Primitive::PointerInteger,
                                Primitive::PointerInteger(42),
                            ),
                        ),
                        "x",
                    )
                    .into()],
                    Return::new(types::Primitive::PointerInteger, Variable::new("x")),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_deconstruct_record() {
            let record_type = types::Record::new(vec![types::Primitive::PointerInteger.into()]);

            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![DeconstructRecord::new(
                        record_type.clone(),
                        Record::new(record_type, vec![Primitive::PointerInteger(42).into()]),
                        0,
                        "x",
                    )
                    .into()],
                    Return::new(types::Primitive::PointerInteger, Variable::new("x")),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_deconstruct_union() {
            let union_type = types::Union::new(vec![types::Primitive::PointerInteger.into()]);

            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![DeconstructUnion::new(
                        union_type.clone(),
                        Union::new(union_type, 0, Primitive::PointerInteger(42)),
                        0,
                        "x",
                    )
                    .into()],
                    Return::new(types::Primitive::PointerInteger, Variable::new("x")),
                ),
                types::Primitive::PointerInteger,
                true,
            ));
        }

        #[test]
        fn compile_compare_and_swap() {
            compile_function_definition(FunctionDefinition::new(
                "f",
                vec![Argument::new(
                    "x",
                    types::Pointer::new(types::Primitive::PointerInteger),
                )],
                Block::new(
                    vec![CompareAndSwap::new(
                        types::Primitive::PointerInteger,
                        Variable::new("x"),
                        Primitive::PointerInteger(0),
                        Primitive::PointerInteger(1),
                        "y",
                    )
                    .into()],
                    Return::new(types::Primitive::Bool, Variable::new("y")),
                ),
                types::Primitive::Bool,
                true,
            ));
        }
    }
}
