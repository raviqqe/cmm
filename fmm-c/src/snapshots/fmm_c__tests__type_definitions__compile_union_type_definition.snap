---
source: fmm-c/src/lib.rs
expression: "compile(&Module::new(vec![VariableDeclaration ::\n                          new(\"x\", types :: Union ::\n                              new(vec !\n                                  [types :: Primitive :: PointerInteger .\n                                   into()]))], vec![], vec![], vec![]))"
---
#include <stdatomic.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
union t_74978b4758aa9d35 {size_t m0;};
extern union t_74978b4758aa9d35 x;
